src/gpu/ffm/trainer.cu

```diff
@@ -0,0 +1,340 @@
+__global__ void wTxKernel(const int *__restrict__ features, const int *__restrict__ fields, const T *__restrict__ values,
+                          const T *__restrict__ r, const int *__restrict__ rowSizes,
+                          T *__restrict__ weightsPtr, T *__restrict__ losses, const bool update, const int *__restrict__ labels = nullptr) {
+  int rowIdx = (blockIdx.x * blockDim.x + threadIdx.x) / cMaxRowSize[0];
+
+  if(rowIdx >= cRows[0]) return;
+
+  int rowSize = rowSizes[rowIdx + 1] - rowSizes[rowIdx];
+
+  int nodeIdx = (blockIdx.x * blockDim.x + threadIdx.x) % cMaxRowSize[0];
+
+  if(nodeIdx >= rowSize) return;
+
+  __shared__ int fieldFeature[MAX_BLOCK_THREADS * 2];
+  __shared__ T vals[MAX_BLOCK_THREADS];
+  __shared__ T scales[MAX_BLOCK_THREADS];
+
+  __shared__ T kappas[MAX_BLOCK_THREADS];
+  __shared__ T expnyts[MAX_BLOCK_THREADS];
+
+  int n1 = rowSizes[rowIdx] + nodeIdx - cBatchOffset[0];
+
+  fieldFeature[threadIdx.x * 2] = fields[n1];
+  fieldFeature[threadIdx.x * 2 + 1] = features[n1];
+  vals[threadIdx.x] = values[n1];
+  scales[rowIdx % MAX_BLOCK_THREADS] = r[rowIdx];
+
+  if(update) {
+    expnyts[rowIdx % MAX_BLOCK_THREADS] = std::exp(-labels[rowIdx] * losses[rowIdx]);
+    kappas[rowIdx % MAX_BLOCK_THREADS] = -labels[rowIdx] * expnyts[rowIdx % MAX_BLOCK_THREADS] / (1 + expnyts[rowIdx % MAX_BLOCK_THREADS]);
+  }
+
+  __syncthreads();
+
+  T loss = 0.0;
+
+  for(int i = 1; n1 + i < rowSizes[rowIdx + 1] - cBatchOffset[0]; i++) {
+    // We cache some of the field:feature:values in shared memory, only as many "nodes" as there are threads
+    // so we know for 100% the initial node will be cached (since we run 1 thread per each starting node)
+    // but the subsequent nodes can be within the same block or they can spill to consequitive blocks
+    // depending on the size of the row and number of threads in a block
+    const int idx1 = fieldFeature[threadIdx.x * 2 + 1] * cAlignFeat[0] +
+        (threadIdx.x + i < MAX_BLOCK_THREADS ? fieldFeature[(threadIdx.x + i) * 2] : fields[n1 + i]) * cAlignFeat[1];
+
+    const int idx2 = (threadIdx.x + i < MAX_BLOCK_THREADS ? fieldFeature[(threadIdx.x + i) * 2 + 1] : features[n1 + i]) * cAlignFeat[0] +
+        fieldFeature[threadIdx.x * 2] * cAlignFeat[1];
+
+    const T v = vals[threadIdx.x] * (threadIdx.x + i < MAX_BLOCK_THREADS ? vals[threadIdx.x + i] : values[n1 + i]) * scales[rowIdx % MAX_BLOCK_THREADS];
+
+    if (update) {
+      for (int d = 0; d < cK[0] * cWeightsOffset[0]; d+=cWeightsOffset[0]) {
+        T w1d = (weightsPtr + idx1)[d];
+        T w2d = (weightsPtr + idx2)[d];
+        const T g1 = cRegLambda[0] * w1d + kappas[rowIdx % MAX_BLOCK_THREADS] * w2d * v;
+        const T g2 = cRegLambda[0] * w2d + kappas[rowIdx % MAX_BLOCK_THREADS] * w1d * v;
+
+        const T w1gdup = (weightsPtr + idx1)[d+1] + g1 * g1;
+        const T w2gdup = (weightsPtr + idx2)[d+1] + g2 * g2;
+
+        (weightsPtr + idx1)[d] -= cLearningRate[0] / std::sqrt(w1gdup) * g1;
```

Comment:
@henrygouk yes to both :-) 

1) definitely different updaters would be great, preferably abstracting this part and encapsulating it in an `Updater` class would be best. Can make an issue with potential candidates

2) for initial release I think it should be ok but in the long run I'm pretty sure indeterministic results will be a no-go, especially for our own use in DAI. Any suggestions how this could be fixed? Serializing it would incur a huge performance hit. Stochastic gradient descent? Not sure how viable that method is? Anything else?